pkg/registry/memory/ns_memory_server.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memory
+
+import (
+	"context"
+
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	"github.com/pkg/errors"
+)
+
+type memoryServiceDiscoveryServer struct {
+	storage *Storage
+}
+
+func (d *memoryServiceDiscoveryServer) FindNetworkService(ctx context.Context, req *registry.FindNetworkServiceRequest) (*registry.FindNetworkServiceResponse, error) {
+	service, ok := d.storage.NetworkServices.Load(req.NetworkServiceName)
+	if !ok {
+		return nil, errors.Errorf("network service %v is not found", req.NetworkServiceName)
+	}
+	NSMs := map[string]*registry.NetworkServiceManager{}
+	var NSEs []*registry.NetworkServiceEndpoint
+	d.storage.NetworkServiceEndpoints.Range(func(_ string, v *registry.NetworkServiceEndpoint) bool {
+		if v.NetworkServiceName == req.NetworkServiceName {
+			NSEs = append(NSEs, v)
+		}
+		return true
+	})
+	for _, nse := range NSEs {
+		nsm, ok := d.storage.NetworkServiceManagers.Load(nse.NetworkServiceManagerName)
+		if ok {
+			NSMs[nsm.Name] = nsm
+		}
+	}
+	return &registry.FindNetworkServiceResponse{
+		Payload: service.GetPayload(),
+		NetworkService: &registry.NetworkService{
+			Name:    service.GetName(),
+			Payload: service.GetPayload(),
+			Matches: service.Matches,
+		},
+		NetworkServiceManagers:  NSMs,
+		NetworkServiceEndpoints: NSEs,
+	}, nil
+}
+
+// NewNetworkServiceDiscoveryServer returns new instance of NetworkServiceDiscoveryServer based on resource client
+func NewNetworkServiceDiscoveryServer(storage *Storage) registry.NetworkServiceDiscoveryServer {
```

Comment:
Why pass in storage as a parameter here, why not just have it created internal to the chain element?

Reply:
Resources should be shared between different registry services. For example, for  NetworkServiceDiscoveryServer, NetworkServeRegistry and etc.

Reply:
Do we have a circumstance where we have a registry server that does not also support Discovery?

Reply:
Actually we have not a  circumstance where we have a registry server that does not also support Discovery. For example in monorepo, `nsmd-k8s`, `proxy-nsmd-k8s`, `floating-registry` have support Discovery.

Reply:
Also, have a look at how this is handled for the [monitor chain element](https://github.com/networkservicemesh/sdk/blob/master/pkg/networkservice/common/monitor/server.go#L49)

Reply:
Not sure that this pattern can be applied for the case of creating three different chain services:
1. https://github.com/networkservicemesh/sdk/blob/master/pkg/registry/core/next/network_service_registry_server.go#L54
2.https://github.com/networkservicemesh/sdk/blob/master/pkg/registry/core/next/nsm_registry_server.go#L72
3. https://github.com/networkservicemesh/sdk/blob/master/pkg/registry/core/next/discovery_server.go#L49

I imagine the use case of these chains as:

```go
storage := new(memory.Storage)
s := tools.NewServer()
// Register discovery service
RegisterNetworkServiceDiscoveryServer(s, 
    next.NewDiscoveryServer(
        ...
        memory.NewDiscoveryServer(storage)))

// Register nse service
RegisterNetworkServiceRegistryServer(s, 
    next.NewNetworkServiceRegistryServer(
        ...
        memory.NewNetworkServiceRegistryServer(storage)))

// Register nsm service
RegisterNsmRegistryServer(s, 
    next.NsmRegistryServer(
        ...
        memory.NsmRegistryServer(storage)))

go func () {
    s.Start() 
}()

```
@edwarnicke  Do I miss something?

Reply:
Hmm... it looks like our registry API has bloated out more than a bit there... how did we wind up with NewNsmRegistryServer in the API? To what purpose?

Reply:
I thought that we can make chains of NSM registry service too. Why it is wrong?

Reply:
I am wondering why the NsmRegistryServer API *exists* at all :)

Reply:
@edwarnicke  I just found a PR when it was separated https://github.com/networkservicemesh/networkservicemesh/pull/805

Reply:
@edwarnicke should we get rid of RegisterNsmRegistryServer?

Reply:
That tells me where it was separated... but why does it exist at all?

Reply:
I think it is related to https://github.com/networkservicemesh/networkservicemesh/issues/799

Reply:
That's still about why it was factored out... why do those RPC calls exist at all?

Reply:
This RPC calls using for the next reasons in monorepo:
1. Registering local NSM by NSMgr on start.
2. Restoring existing endpoints in case of restart local NSMgr.

Reply:
https://github.com/networkservicemesh/networkservicemesh/blob/master/controlplane/pkg/nsmd/nsmd.go#L525

Reply:
How it was added into API https://github.com/networkservicemesh/api/commit/0623273c4e70ea80e64087a7a9ff217c9dbdb358

Reply:
@edwarnicke not quite understand what you want to do with `NewNsmRegistryServer`. Please share your thoughts.